Print memory usage (p/P) only on change

ght opened the pull request:
No longer spam the memory usage constantly. This makes the output much more useful by allowing to observe each change in memory usage, since the values now no longer get pushed up and scroll out of view rapidly.

---

Print used memory (and change since last print) in addition to free memory.

Example output:
```
Mem: 4952 used (-248) 9384 free
Mem: 5960 used (+1008) 8376 free
Mem: 5992 used ( +32) 8344 free
Mem: 5720 used (-272) 8616 free
Mem: 5728 used (  +8) 8608 free
Mem: 5880 used (+152) 8456 free
Mem: 5728 used (-152) 8608 free
```

NeroBurner commented:
Thanks for the welcome improvement!

One possible Problem: I can set the lvgl total memory up to debug out of memory  issues. This means the free memory theoretically can be negative. Using size_t would then lead to huuuuuge positive numbers. This is why I initially used int64_t to print the free memory

ght commented:
Good catch, thank you! :)

It actually didn't really manifest as a bug, since I had `%d` in the `printf`, so the unsigned `size_t` got reinterpreted as signed when printing and the expected values were printed (I tested by setting `pinetime_lvgl_memory` to a low enough value to cause negative numbers being printed). Still not correct and an easy fix though.

I dug into the LVGL source to see what type `mem_mon.free_size` actually has (`uint32_t`, [see here](https://github.com/lvgl/lvgl/blob/v7.11.0/src/lv_misc/lv_mem.h#L45)) and adjusted the types accordingly.

I also removed the `int64_t` conversion, since I cannot see how it would be necessary. I tried to blame the line to see where it came from, but that just unhelpfully gives [the initial commit for InfiniSim](https://github.com/InfiniTimeOrg/InfiniSim/blame/main/main.cpp#L838). If it is indeed needed, let me know and I'll put it back in :)

NeroBurner commented on main.cpp:
```diff
@@ -832,11 +832,18 @@ class Framework {
-        printf("actual free_size = %d\n", int64_t(mem_mon.free_size) - (LV_MEM_SIZE - 14U*1024U));
+        if (mem_mon.free_size != mem_mon_last_free_size) {
+          uint32_t pinetime_lvgl_memory = 14U*1024U; // 14KiB is the LVGL memory size used in InfiniTime
+          uint32_t mem_used = LV_MEM_SIZE - int64_t(mem_mon.free_size);
```
I think this `int64_t` cast isn't needed anymore if we assume, that free_size can't be greater than `LV_MEM_SIZE`. Furthermore in the case `free_size` actually gets greater than `LV_MEM_SIZE` the cast to `int64_t`, then compute the diff, and then cast back to `uint32_t` will still lead to a negative number casted to an unsigned number, resulting in a huge number

i propose to add an `assert(LV_MEM_SIZE >= mem_mon.free_size)` statement to at least check for that case in debug builds. And to remove the `int64_t` cast here

ght replied:
I removed the cast, but decided against adding the proposed `assert`, since [LVGL allocates only `LV_MEM_SIZE` bytes](https://github.com/lvgl/lvgl/blob/v7.11.0/src/lv_misc/lv_mem.c#L120) with `LV_MEM_CUSTOM=0` and `LV_MEM_ADR=0`, which InfiniTime uses.

So `lv_mem_monitor` reporting a `free_size` larger than `LV_MEM_SIZE` would be a major bug in LVGL and if we were to guard against that, the `assert` should go higher up so it is not only tripped when memory debugging is toggled :)

NeroBurner commented on main.cpp:
```diff
@@ -832,11 +832,18 @@ class Framework {
         lv_mem_monitor(&mem_mon);
-        printf("actual free_size = %d\n", int64_t(mem_mon.free_size) - (LV_MEM_SIZE - 14U*1024U));
+        if (mem_mon.free_size != mem_mon_last_free_size) {
+          uint32_t pinetime_lvgl_memory = 14U*1024U; // 14KiB is the LVGL memory size used in InfiniTime
```
the compiler most probably does this, but please make this constant a `constexpr` to explicitly mark it as compile time constant

NeroBurner reviewed (changes requested):
last little cleanups, then we're good to go :)

ght commented:
Changes:
- Add `max_used` peak memory usage
- Add labels in the output (slightly more verbose, but much clearer)
- Add `constexpr` to `pinetime_lvgl_memory`
- Remove unnecessary `int64_t` cast
- Change "free" to "budget left" to clarify it shows mem that would be
  free on the real hardware, making situations where it goes negative
  less confusing
- Change field widths so 5 digit values no longer jump out of alignment

Example output:
```
Mem:  6720 used (change:  -248, peak:  7928) 7616 budget left
Mem:  6832 used (change:  +112, peak:  7928) 7504 budget left
Mem:  6584 used (change:  -248, peak:  7928) 7752 budget left
Mem: 10856 used (change: +4272, peak: 10848) 3480 budget left
Mem: 10888 used (change:   +32, peak: 10880) 3448 budget left
Mem: 10616 used (change:  -272, peak: 10880) 3720 budget left
Mem:  6704 used (change: -3912, peak: 10880) 7632 budget left
Mem:  6808 used (change:  +104, peak: 10880) 7528 budget left
```

NeroBurner reviewed (approved):
Looking really good! Thank you very much!
